tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -0,0 +1,594 @@
+    *
+    * @param decl an operator declaration
+    * @param inScopeEx an expression to chain in the bottom of let-definition, it should have been translated
+    * @return the translated let-in expression with inScopeEx attached
+    */
+  def apply(decl: TlaOperDecl, inScopeEx: EtcExpr): EtcExpr = {
+    val paramsAndDoms = formalParamsToTypeVars(0, decl.formalParams).map {
+      case (paramName, paramType) =>
+        (paramName, mkConst(BlameRef(decl.body.ID), SetT1(paramType)) )
+    }
+
+    if (decl.name.startsWith("TypeAssumptions")) {
+      // the special form for type annotations over state variables and constants
+      parseTypeAssumptions(decl, inScopeEx)
+    } else {
+      // The type of the lambda is what we want to see as the type of the declaration.
+      // There are two cases: (1) the definition body contains a type annotation, and (2) no type annotation
+      decl.body match {
+        case OperEx(TypingOper.withType, ValEx(TlaStr(typeText)), body) =>
+          // case 1: the definition body contains a type annotation
+          val parsedType =
+            try {
+              type1Parser(typeText)
+            } catch {
+              case e: Type1ParseError =>
+                throw new TypingInputException(s"Parser error in type annotation of ${decl.name}: ${e.msg}")
+            }
+
+          val lambda = mkAbs(ExactRef(body.ID), this (body), paramsAndDoms: _*)
+          val letIn = mkLet(BlameRef(body.ID), decl.name, lambda, inScopeEx)
+          mkTypeDecl(ExactRef(decl.body.ID), decl.name, parsedType, letIn)
+
+        case _ =>
+          // case 2: no type annotation
+          val lambda = mkAbs(ExactRef(decl.body.ID), this (decl.body), paramsAndDoms: _*)
+          mkLet(BlameRef(decl.body.ID), decl.name, lambda, inScopeEx)
+      }
+    }
+  }
+
+  // parse the type annotations inside TypeAssumptions
+  private def parseTypeAssumptions(decl: TlaOperDecl, inScopeEx: EtcExpr): EtcExpr = {
+    decl.body match {
+      case OperEx(TlaBoolOper.and, args @ _*) =>
+        val annotations =
+          args.collect { case OperEx(TypingOper.assumeType, NameEx(name), ValEx(TlaStr(typeText))) =>  (name, typeText) }
+        if (annotations.length != args.length) {
+          throw new TypingInputException(s"""Error in ${decl.name}: Expected AssumeType(varName, "typeString")""")
+        } else {
+          annotations.foldRight(inScopeEx) {
+            case ((name, typeText), terminal) =>
+              try {
+                val tt = type1Parser(typeText)
+                mkTypeDecl(ExactRef(decl.body.ID), name, tt, terminal)
+              } catch {
+                case e: Type1ParseError =>
+                  throw new TypingInputException(s"Parser error in type annotation of $name: ${e.msg}")
+              }
+          }
+        }
+
+      case _ =>
+        throw new TypingInputException(s"""Error in ${decl.name}: Expected /\ AssumeType(varName, "typeString") /\ ...""")
```

Comment:
I could see users getting caught up here pretty easily.

Ideally it would link to the documentation on this convention.